### PR TITLE
[SPARK-46547][SS] Swallow non-fatal exception in maintenance task to avoid deadlock between maintenance thread and streaming aggregation operator

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Swallow non-fatal exception in maintenance task to avoid deadlock between maintenance thread and streaming aggregation operator

### Why are the changes needed?
This change fixes a race condition that causes a deadlock between the task thread and the maintenance thread. This is primarily only possible with the streaming aggregation operator. In this case, we use 2 physical operators - `StateStoreRestoreExec` and `StateStoreSaveExec`. The first one opens the store in read-only mode and the 2nd one does the actual commit.

However, the following sequence of events creates an issue
1. Task thread runs the `StateStoreRestoreExec` and gets the store instance and thereby the DB instance lock
2. Maintenance thread fails with an error for some reason
3. Maintenance thread takes the `loadedProviders` lock and tries to call `close` on all the loaded providers
4. Task thread tries to execute the StateStoreRDD for the `StateStoreSaveExec` operator and tries to acquire the `loadedProviders` lock which is held by the thread above

So basically if the maintenance thread is interleaved between the `restore/save` operations, there is a deadlock condition based on the `loadedProviders` lock and the DB instance lock.

The fix proposes to simply release the resources at the end of the `StateStoreRestoreExec` operator (note that `abort` for `ReadStateStore` is likely a misnomer - but we choose to follow the already provided API in this case)

Relevant Logs:
Link - https://github.com/anishshri-db/spark/actions/runs/7356847259/job/20027577445?pr=4
```
2023-12-27T09:59:02.6362466Z 09:59:02.635 WARN org.apache.spark.sql.execution.streaming.state.StateStore: Error in maintenanceThreadPool
2023-12-27T09:59:02.6365616Z java.io.FileNotFoundException: File file:/home/runner/work/spark/spark/target/tmp/spark-8ef51f34-b9de-48f2-b8df-07e14599b4c9/state/0/1 does not exist
2023-12-27T09:59:02.6367861Z 	at org.apache.hadoop.fs.RawLocalFileSystem.listStatus(RawLocalFileSystem.java:733)
2023-12-27T09:59:02.6369383Z 	at org.apache.hadoop.fs.DelegateToFileSystem.listStatus(DelegateToFileSystem.java:177)
2023-12-27T09:59:02.6370693Z 	at org.apache.hadoop.fs.ChecksumFs.listStatus(ChecksumFs.java:571)
2023-12-27T09:59:02.6371781Z 	at org.apache.hadoop.fs.FileContext$Util$1.next(FileContext.java:1940)
2023-12-27T09:59:02.6372876Z 	at org.apache.hadoop.fs.FileContext$Util$1.next(FileContext.java:1936)
2023-12-27T09:59:02.6373967Z 	at org.apache.hadoop.fs.FSLinkResolver.resolve(FSLinkResolver.java:90)
2023-12-27T09:59:02.6375104Z 	at org.apache.hadoop.fs.FileContext$Util.listStatus(FileContext.java:1942)
2023-12-27T09:59:02.6376676Z 09:59:02.636 WARN org.apache.spark.sql.execution.streaming.state.StateStore: Error running maintenance thread
2023-12-27T09:59:02.6379079Z java.io.FileNotFoundException: File file:/home/runner/work/spark/spark/target/tmp/spark-8ef51f34-b9de-48f2-b8df-07e14599b4c9/state/0/1 does not exist
2023-12-27T09:59:02.6381083Z 	at org.apache.hadoop.fs.RawLocalFileSystem.listStatus(RawLocalFileSystem.java:733)
2023-12-27T09:59:02.6382490Z 	at org.apache.hadoop.fs.DelegateToFileSystem.listStatus(DelegateToFileSystem.java:177)
2023-12-27T09:59:02.6383816Z 	at org.apache.hadoop.fs.ChecksumFs.listStatus(ChecksumFs.java:571)
2023-12-27T09:59:02.6384875Z 	at org.apache.hadoop.fs.FileContext$Util$1.next(FileContext.java:1940)
2023-12-27T09:59:02.6386294Z 	at org.apache.hadoop.fs.FileContext$Util$1.next(FileContext.java:1936)
2023-12-27T09:59:02.6387439Z 	at org.apache.hadoop.fs.FSLinkResolver.resolve(FSLinkResolver.java:90)
2023-12-27T09:59:02.6388674Z 	at org.apache.hadoop.fs.FileContext$Util.listStatus(FileContext.java:1942)
...
2023-12-27T10:01:02.4292831Z [0m[[0m[0minfo[0m] [0m[0m[31m- changing schema of state when restarting query - state format version 2 (RocksDBStateStore) *** FAILED *** (2 minutes)[0m[0m
2023-12-27T10:01:02.4295311Z [0m[[0m[0minfo[0m] [0m[0m[31m  Timed out waiting for stream: The code passed to failAfter did not complete within 120 seconds.[0m[0m
2023-12-27T10:01:02.4297271Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.base/java.lang.Thread.getStackTrace(Thread.java:1619)[0m[0m
2023-12-27T10:01:02.4299084Z [0m[[0m[0minfo[0m] [0m[0m[31m  	org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:277)[0m[0m
2023-12-27T10:01:02.4300948Z [0m[[0m[0minfo[0m] [0m[0m[31m  	org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)[0m[0m
...
2023-12-27T10:01:02.6474472Z 10:01:02.646 WARN org.apache.spark.sql.execution.streaming.state.RocksDB StateStoreId(opId=0,partId=0,name=default): Error closing RocksDB
2023-12-27T10:01:02.6482792Z org.apache.spark.SparkException: [CANNOT_LOAD_STATE_STORE.UNRELEASED_THREAD_ERROR] An error occurred during loading state. StateStoreId(opId=0,partId=0,name=default): RocksDB instance could not be acquired by [ThreadId: Some(1858)] as it was not released by [ThreadId: Some(3835), task: partition 0.0 in stage 513.0, TID 1369] after 120009 ms.
2023-12-27T10:01:02.6488483Z Thread holding the lock has trace: app//org.apache.spark.sql.execution.streaming.state.StateStore$.getStateStoreProvider(StateStore.scala:577)
2023-12-27T10:01:02.6490896Z app//org.apache.spark.sql.execution.streaming.state.StateStore$.get(StateStore.scala:565)
2023-12-27T10:01:02.6493072Z app//org.apache.spark.sql.execution.streaming.state.StateStoreRDD.compute(StateStoreRDD.scala:128)
2023-12-27T10:01:02.6494915Z app//org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
2023-12-27T10:01:02.6496232Z app//org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
2023-12-27T10:01:02.6497655Z app//org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
2023-12-27T10:01:02.6499153Z app//org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
2023-12-27T10:01:02.6556758Z 10:01:02.654 WARN org.apache.spark.scheduler.TaskSetManager: Lost task 0.0 in stage 513.0 (TID 1369) (localhost executor driver): TaskKilled (Stage cancelled: [SPARK_JOB_CANCELLED] Job 260 cancelled part of cancelled job group cf26288c-0158-48ce-8a86-00a596dd45d8 SQLSTATE: XXKDA)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests
```
[info] Run completed in 6 minutes, 20 seconds.
[info] Total number of tests run: 80
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 80, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
Yes

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @HeartSaVioR - PTAL whenever you get a chance, thx !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 1]
```diff
@@ -434,22 +434,26 @@ case class StateStoreRestoreExec(
...
+          val key = stateManager.getKey(row.asInstanceOf[UnsafeRow])
+          val restoredRow = stateManager.get(store, key)
+          val outputRows = Option(restoredRow).toSeq :+ row
+          numOutputRows += outputRows.size
+          outputRows
         }
+      }
+      // SPARK-46547 - Release any locks/resources if required, to prevent
+      // deadlocks with the maintenance thread.
+      store.abort()
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:455]: @HeartSaVioR - we could probably make this more light-weight - in order to just release the instance lock and keep the loaded version intact. but then we probably need to add another API on the store or pass an argument to `abort` ? Thoughts ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 2]
```diff
@@ -434,22 +434,26 @@ case class StateStoreRestoreExec(
...
-          store.iterator().map(_.value)
-        } else {
-          iter.flatMap { row =>
-            val key = stateManager.getKey(row.asInstanceOf[UnsafeRow])
-            val restoredRow = stateManager.get(store, key)
-            val outputRows = Option(restoredRow).toSeq :+ row
-            numOutputRows += outputRows.size
-            outputRows
-          }
+      val hasInput = iter.hasNext
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:437]: Is this spurious change? Seems like indentation changed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala (hunk 1 above)
Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:455]: Btw, why was this not an issue before? When are these locks/resources released before this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:437]: Yea just fixed the indentation

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:455]: I believe this issue always existed. In the common case, we won't see this that often though. Basically the restore is followed by the save operator which opens the db instance in read-write mode. At the end of the save operator, we would always release the instance lock. For the `ReadStateStore` invocations though, there is no such functionality (we only abort on task failure) - so the interleaving of the maintenance thread error case (also likely rare) and the execution of StateStoreRDD for the `save` operator causes this deadlock

**Comment by HeartSaVioR** [comment]
I believe the real thing here is that the failure of maintenance task is hammering all active state store providers, effectively impacting to all stateful tasks on the executor.

Let's look back what we do in maintenance task. Mostly we do snapshotting and cleaning up orphaned files. If we suppose the task fails, would the state store (provider) be impacted? From what I understand, no, it is not impacted.

This is reflected in the HDFS backed state store provider. If we look at maintenance task in HDFS backed state store provider, it swallows non-fatal exception. If we agree that the failure of maintenance task in RocksDB state store provider does not impact the actual state store (provider), we can do the same to RocksDB state store provider.

**Comment by HeartSaVioR** [comment]
That said, what we really need to fix is the behavior when maintenance task fails. A single streaming query using faulty state store provider implementation can lead every other stateful queries to fail in the same executor. While this is really something we have to fix (maybe only close the state store provider for the faulty one), but as I commented earlier, if we do not think the failure of maintenance task in RocksDB state store provider impacts the actual state store (provider), we should just swallow the exception.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 3]
```diff
@@ -84,10 +84,11 @@ trait ReadStateStore {
 
   /**
    * Clean up the resource.
-   *
+   * @param releaseOnly - if true, only release the instance lock and do not run the full abort
+   *                    sequence.
    * The method name is to respect backward compatibility on [[StateStore]].
    */
-  def abort(): Unit
+  def abort(releaseOnly: Boolean = false): Unit
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: Reverted and moved to the exception handling approach. PTAL

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:1]: nit: revert the change, previous indentation seems to be right

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 4]
```diff
@@ -452,6 +452,7 @@ class RocksDB(
    * Drop uncommitted changes, and roll back to previous version.
    */
   def rollback(): Unit = {
+    acquire()
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:455]: nit: This is now a side-effect change. Do we want to keep this or just remove this? I feel like it's safer to add acquire() here, but once you update the PR title and description to reflect the new direction, this change is beyond the scope.

**Review by HeartSaVioR — COMMENTED** [review]
Only nits. Could you please update the PR title and description to reflect the new direction? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:1]: If you look at the indent pattern - the braces seem off. It seems this pattern is used through this file though. Ok - will just update to original

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:1]: Actually looked at other places - it does seem that they are fine. Could you please check the pattern once again ?

```
    child.execute().mapPartitionsWithReadStateStore(
      getStateInfo,
      keyExpressions.toStructType,
      stateManager.getStateValueSchema,
      numColsPrefixKey = 0,
      session.sessionState,
      Some(session.streams.stateStoreCoordinator)) { case (store, iter) =>
``

2 spaces are already added in the beginning. Do we need 2 more for the next line statement ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 4 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:455]: This is actually not exactly related to this change. But we had missed this call for this function. Do you prefer to create this in a separate change ?

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala]: Discussed with @HeartSaVioR - this is broken. but will fix this in a separate PR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:1]: Same here - will fix in a separate PR

**Review by HeartSaVioR — APPROVED** [review]
+1 Nice RCA and thanks for the concise fix!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 5]
```diff
@@ -233,7 +235,12 @@ private[sql] class RocksDBStateStoreProvider
   }
 
   override def doMaintenance(): Unit = {
-    rocksDB.doMaintenance()
+    try {
+      rocksDB.doMaintenance()
+    } catch {
+      case NonFatal(ex) =>
+        logWarning(s"Error performing maintenance operations with exception=$ex")
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:242]: Please change this to include full stacktrace `logWarning("...", ex)`

Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:242]: Also make it explicit that this error is ignored in the log.

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Done - updated the PR

**Review by rangadi — APPROVED** [review]
Could you update the description now that fix has changed after @HeartSaVioR's comments? 
I didn't fully grok the issue or the fix. Overall, avoiding throwing seems like a work around not to trigger the bug. I am sure both of you have discussed it. 
LGTM.

**Comment by HeartSaVioR** [comment]
@rangadi 

Please read through my comments in above. Here are links for you:

* https://github.com/apache/spark/pull/44542#issuecomment-1882621196
* https://github.com/apache/spark/pull/44542#issuecomment-1882649328

**Comment by HeartSaVioR** [comment]
https://github.com/anishshri-db/spark/actions/runs/7471692378/job/20332452975
Failing module is unrelated.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.
